pkg/restore/client.go

```diff
@@ -457,6 +468,154 @@ func (rc *Client) ValidateChecksum(
+		}
+		for _, l := range s.GetLabels() {
+			if l.GetKey() == restoreLabelKey && l.GetValue() == restoreLabelValue {
+				rc.restoreStores = append(rc.restoreStores, s.GetId())
+				break
+			}
+		}
+	}
+	log.Info("load restore stores", zap.Uint64s("store-ids", rc.restoreStores))
+	return nil
+}
+
+// ResetRestoreLabels removes the exclusive labels of the restore stores.
+func (rc *Client) ResetRestoreLabels(ctx context.Context) error {
+	if !rc.isOnline {
+		return nil
+	}
+	log.Info("start reseting store labels")
+	return rc.toolClient.SetStoresLabel(ctx, rc.restoreStores, restoreLabelKey, "")
+}
+
+// SetupPlacementRules sets rules for the tables' regions.
+func (rc *Client) SetupPlacementRules(ctx context.Context, tables []*model.TableInfo) error {
+	if !rc.isOnline || len(rc.restoreStores) == 0 {
+		return nil
+	}
+	log.Info("start setting placement rules")
+	rule, err := rc.toolClient.GetPlacementRule(ctx, "pd", "default")
+	if err != nil {
+		return err
+	}
+	rule.Index = 100
+	rule.Override = true
+	rule.LabelConstraints = append(rule.LabelConstraints, placement.LabelConstraint{
+		Key:    restoreLabelKey,
+		Op:     "in",
+		Values: []string{restoreLabelValue},
+	})
+	for _, t := range tables {
+		rule.ID = rc.getRuleID(t.ID)
+		rule.StartKeyHex = hex.EncodeToString(codec.EncodeBytes([]byte{}, tablecodec.EncodeTablePrefix(t.ID)))
+		rule.EndKeyHex = hex.EncodeToString(codec.EncodeBytes([]byte{}, tablecodec.EncodeTablePrefix(t.ID+1)))
+		err = rc.toolClient.SetPlacementRule(ctx, rule)
+		if err != nil {
+			return err
+		}
+	}
+	log.Info("finish setting placement rules")
+	return nil
+}
+
+// WaitPlacementSchedule waits PD to move tables to restore stores.
+func (rc *Client) WaitPlacementSchedule(ctx context.Context, tables []*model.TableInfo) error {
+	if !rc.isOnline || len(rc.restoreStores) == 0 {
+		return nil
+	}
+	log.Info("start waiting placement schedule")
+	ticker := time.NewTicker(time.Second * 10)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			ok, progress, err := rc.checkRegions(ctx, tables)
+			if err != nil {
+				return err
+			}
+			if ok {
+				log.Info("finish waiting placement schedule")
+				return nil
+			}
+			log.Info("placement schedule progress: " + progress)
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+func (rc *Client) checkRegions(ctx context.Context, tables []*model.TableInfo) (bool, string, error) {
+	for i, t := range tables {
+		start := codec.EncodeBytes([]byte{}, tablecodec.EncodeTablePrefix(t.ID))
+		end := codec.EncodeBytes([]byte{}, tablecodec.EncodeTablePrefix(t.ID+1))
+		ok, regionProgress, err := rc.checkRange(ctx, start, end)
+		if err != nil {
+			return false, "", err
+		}
+		if !ok {
+			return false, fmt.Sprintf("table %v/%v, %s", i, len(tables), regionProgress), nil
+		}
+	}
+	return true, "", nil
+}
+
+func (rc *Client) checkRange(ctx context.Context, start, end []byte) (bool, string, error) {
+	regions, err := rc.toolClient.ScanRegions(ctx, start, end, -1)
```

Comment:
I think it should not. It only scans regions to load new tables, which are basically 1 region per table, because the check happends before loading data or region split.